lib/galaxy/datatypes/binary.py

```diff
@@ -462,6 +473,58 @@ def to_archive(self, trans, dataset, name=""):
         file_paths.append(dataset.metadata.bam_index.file_name)
         return zip(file_paths, rel_paths)
 
+    def get_chunk( self, trans, dataset, offset=0, ck_size=None ):
+        index_file = dataset.metadata.bam_index
+        with pysam.AlignmentFile( dataset.file_name, "rb", index_filename=index_file.file_name ) as bamfile:
+            ck_size = 300  # 300 lines
+            ck_data = ""
+            last_read = 0
+            header_line_count = 0
+            if offset == 0:
+                ck_data = bamfile.text
+                header_line_count = bamfile.text.count('\n')
+            else:
+                bamfile.seek( offset )
+            for line_number, alignment in enumerate( bamfile ) :
+                # return only Header lines if 'header_line_count' exceeds 'ck_size'
+                # FIXME: Can be problematic if bam has million lines of header
+                if ( line_number + header_line_count ) > ck_size:
+                    break
+                else:
+                    bamline = alignment.tostring( bamfile )
+                    # Galaxy display each tag as separate column because 'tostring()' funcition put spaces in between each tag of tags column.
+                    # Below code will remove spaces between each tag.
+                    bamline_modified = ('\t').join( bamline.split()[:11] + [ ('').join(bamline.split()[11:]) ] )
+                    ck_data = ck_data + "\n" + bamline_modified
+                last_read = bamfile.tell()
```

Comment:
if i put the ` last_read = bamfile.tell()` , I am missing one line form bamfile for every chunk. Because, I am breaking the loop only after enumerating.